Modules/unicodedata.c

```diff
@@ -1253,6 +1289,109 @@ unicodedata_UCD_lookup_impl(PyObject *self, const char *name,
     return PyUnicode_FromOrdinal(code);
 }
 
+typedef struct {
+    PyObject_HEAD
+    PyObject* str;
+    Py_ssize_t pos;
+    Py_ssize_t end;
+} GraphemeClusterIterator;
+
+static void
+GCI_dealloc(GraphemeClusterIterator *it)
+{
+    PyObject_GC_UnTrack(it);
+    Py_DECREF(it->str);
+    PyObject_GC_Del(it);
+}
+
+static int
+GCI_traverse(GraphemeClusterIterator *it, visitproc visit, void *arg) {
+    Py_VISIT(it->str);
+    return 0;
+}
+
+static int
+GCI_clear(GraphemeClusterIterator *self)
+{
+    Py_CLEAR(self->str);
+    return 0;
+}
+
+#include "grapheme_cluster_break_automaton.h"
+
+static PyObject *
+GCI_iternext(GraphemeClusterIterator *self)
+{
+    int kind = PyUnicode_KIND(self->str);
+    void *pstr = PyUnicode_DATA(self->str);
+    if (self->pos == self->end) {
+        return NULL;
+    }
+
+    int start = self->pos;
+    GCBState s = STATE_sot;
+    while (1) {
+        if (self->pos == self->end) {
+            return PyUnicode_Substring(self->str, start, self->pos);
+        }
+        Py_UCS4 chr = PyUnicode_READ(kind, pstr, self->pos);
+        int prop = _getrecord_ex(chr)->grapheme_cluster_break;
+        s = GRAPH_CLUSTER_AUTOMATON[s][prop];
+        if (s == STATE_BREAK) {
+            return PyUnicode_Substring(self->str, start, self->pos);
+        }
+        ++self->pos;
+    }
+}
+
+static PyTypeObject GraphemeClusterIteratorType = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    .tp_name = "unicodedata.GraphemeClusterIterator",
+    .tp_basicsize = sizeof(GraphemeClusterIterator),
+    .tp_dealloc = (destructor)GCI_dealloc,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_doc = "Internal grapheme cluster iterator object.",
+    .tp_iter = PyObject_SelfIter,
+    .tp_iternext = (iternextfunc)GCI_iternext,
+    .tp_traverse = (traverseproc)GCI_traverse,
+    .tp_clear = (inquiry)GCI_clear
+};
+
+/*[clinic input]
+unicodedata.UCD.iter_graphemes
+
+    self: self
+    unistr: unicode
+    start: int = 0
+    end: Py_ssize_t(c_default="PY_SSIZE_T_MAX - 1") = sys.maxsize
```

Comment:
It should be PY_SSIZE_T_MAX.

Although I am not sure that the `end` parameter is needed. The user can simply stop iteration at any time.